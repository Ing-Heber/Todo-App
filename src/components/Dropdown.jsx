import {useState, useRef, useEffect} from "react";
import PropTypes from "prop-types";

export const Dropdown = ({btnClasses, btnText, dropdownId, dropdownType, handleShowTodosBy, items}) => {
    const [isOpen, setIsOpen] = useState(false);
    const [selected, setSelected] = useState(0);
    const [focusedIndex, setFocusedIndex] = useState(-1);
    const dropdownRef = useRef(null);
    const buttonRef = useRef(null);
    const itemRefs = useRef([]);

    const toggleDropdown = () => {
        setIsOpen(!isOpen);
        setFocusedIndex(-1);
    };

    const closeDropdown = () => {
        setIsOpen(false);
        setFocusedIndex(-1);
        buttonRef.current?.focus();
    };

    const handleSelect = (index) => {
        setSelected(index);
        handleShowTodosBy(items[index].value);
        closeDropdown();
    };

    const handleKeyDown = (e) => {
        switch (e.key) {
            case 'Escape':
                e.preventDefault();
                closeDropdown();
                break;
            case 'Enter':
            case ' ':
                e.preventDefault();
                if (isOpen && focusedIndex >= 0) {
                    handleSelect(focusedIndex);
                } else {
                    toggleDropdown();
                }
                break;
            case 'ArrowDown':
                e.preventDefault();
                if (!isOpen) {
                    setIsOpen(true);
                    setFocusedIndex(0);
                } else {
                    setFocusedIndex((prev) =>
                        prev < items.length - 1 ? prev + 1 : 0
                    );
                }
                break;
            case 'ArrowUp':
                e.preventDefault();
                if (isOpen) {
                    setFocusedIndex((prev) =>
                        prev > 0 ? prev - 1 : items.length - 1
                    );
                }
                break;
            case 'Tab':
                if (isOpen) {
                    closeDropdown();
                }
                break;
            default:
                break;
        }
    };

    const handleItemKeyDown = (e, index) => {
        switch (e.key) {
            case 'ArrowUp':
                if (index <= 0) setFocusedIndex(items.length - 1);
                setFocusedIndex(index - 1);
                break;

            case 'ArrowDown':
                if (index >= items.length - 1) setFocusedIndex(0);
                setFocusedIndex(index + 1);
                break;

            case 'Enter':
            case ' ':
                e.preventDefault();
                handleSelect(index);
                break;
            case 'Escape':
                e.preventDefault();
                closeDropdown();
                break;
            default:
                break;
        }
    };

    useEffect(() => {
        const handleClickOutside = (event) => {
            if (dropdownRef.current && !dropdownRef.current.contains(event.target)) {
                closeDropdown();
            }
        };

        if (isOpen) {
            document.addEventListener('mousedown', handleClickOutside);
        }

        return () => {
            document.removeEventListener('mousedown', handleClickOutside);
        };
    }, [isOpen]);

    useEffect(() => {
        if (isOpen && focusedIndex >= 0 && itemRefs.current[focusedIndex]) {
            itemRefs.current[focusedIndex].focus();
        }
    }, [focusedIndex, isOpen]);

    return (
        <div className="dropdown" ref={dropdownRef} aria-label={`${dropdownType} dropdown`}>
            <button
                ref={buttonRef}
                className={`btn__text--primary dropdown__toggle ${btnClasses}`}
                onClick={toggleDropdown}
                onKeyDown={handleKeyDown}
                aria-expanded={isOpen}
                aria-haspopup="listbox"
                aria-controls={dropdownId}
                aria-label={`${btnText}, current selection: ${items[selected].label}`}
            >
                <span className="d-md-block">{btnText}</span>
                <p className="count small">{items[selected].label}</p>
            </button>

            {isOpen && (
                <div
                    id={dropdownId}
                    className="dropdown__container show"
                    role="listbox"
                    aria-labelledby={dropdownId}
                >
                    {items.map((item, index) => (
                        <button
                            key={`${item.value}-${index}`}
                            ref={(el) => (itemRefs.current[index] = el)}
                            className={`dropdown__item ${selected === index ? 'active' : ''} ${
                                focusedIndex === index ? 'focused' : ''
                            }`}
                            onClick={() => handleSelect(index)}
                            onKeyDown={(e) => handleItemKeyDown(e, index)}
                            role="option"
                            aria-selected={selected === index}
                            tabIndex={0}
                        >
                            {item.label}
                        </button>
                    ))}
                </div>
            )}
        </div>
    );
};

Dropdown.propTypes = {
    btnClasses: PropTypes.string,
    btnText: PropTypes.string.isRequired,
    dropdownId: PropTypes.string.isRequired,
    dropdownType: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string.isRequired,
            value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        })
    ).isRequired,
    handleShowTodosBy: PropTypes.func.isRequired,
};
