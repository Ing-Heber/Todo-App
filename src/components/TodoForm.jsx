import PropTypes from 'prop-types'
import { useForm } from '../hooks/useForm';

export const TodoForm = ({ onNewTodo }) => {

  const { description, onInputChange, onResetForm } = useForm({
    description: '',
  });


  const onSubmit = (event) => {
    const newTodo = {
      id: new Date().getTime() * 3,
      description: description,
      completed: false
    }

    event.preventDefault();
    if (description.trim().length < 3) return;

    onNewTodo(newTodo);
    onResetForm();
  }

  return (
    <form className='inline-form' onSubmit={onSubmit}>

      <input className='input__text' name='description' type="text" value={description} onChange={onInputChange} placeholder='Add Todo' />

      <button aria-label="Add Todo" type='submit' className='btn__solid--primary'>
        <span className='d-none d-md-block'>Add Todo</span>
        <img src='/assets/plus.svg' alt="Plus Image" width='16px' height='auto' />
      </button>

    </form>
  )
};

TodoForm.propTypes = {
  onNewTodo: PropTypes.func
}