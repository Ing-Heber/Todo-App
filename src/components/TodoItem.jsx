import PropTypes from 'prop-types'
import { TrashIcon } from './TrashIcon'
import { useTodoItem } from '../hooks/useTodoItem'

export const TodoItem = ({ id, description, completed, toggleTodo, deleteTodo }) => {

    const {
        cardAnimation,
        descriptionStyle,
        onToggle,
        onDelete
    } = useTodoItem({ id, completed, toggleTodo, deleteTodo });

    return (
        <div className={`todo__card animate__animated ${cardAnimation}`}>
            <div className='card__content d-flex align-items-center'>
                <input className='input__checkbox' type="checkbox" name={description} id={id} checked={completed} onChange={onToggle} />
                <label htmlFor={id} className={`card__description ${descriptionStyle}`} >{description}</label>
            </div>
            <button aria-label='delete button' className='btn--delete' onClick={onDelete}>
                <TrashIcon />
            </button>
        </div>
    )
}

TodoItem.propTypes = {
    id: PropTypes.number,
    description: PropTypes.string,
    completed: PropTypes.bool,
    toggleTodo: PropTypes.func,
    deleteTodo: PropTypes.func
}

