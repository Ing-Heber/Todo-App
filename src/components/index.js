export * from './TodoEmpty'
export * from './TodoForm'
export * from './TodoItem'
export * from './TodoList'
export * from './TrashIcon'