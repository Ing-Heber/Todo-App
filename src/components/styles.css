/* Elements */

button {
    color: var(--gray-100);
    cursor: pointer;
    font-family: Inter;
    font-size: 14px;
    font-style: normal;
    font-weight: 700;
    line-height: 140%;
}


button.btn__solid--primary {
    align-items: center;
    background: var(--blue-dark);
    border-radius: 8px;
    border: none;
    display: flex;
    gap: 8px;
    justify-content: center;
    padding: 16px;
}

button.btn__solid--primary:hover,
button.btn__solid--primary:active,
button.btn__solid--primary:focus-visible {
    background: var(--blue);
    outline: none;
}


button.btn--delete {
    background: none;
    border: none;
    border-radius: 4px;
    padding: 5px 5.522px 5px 6px;
}

button.btn--delete:hover,
button.btn--delete:active,
button.btn--delete:focus-visible {
    background: var(--gray-400);
    outline: none;
}

:where(button.btn--delete:hover,
    button.btn--delete:active,
    button.btn--delete:focus-visible) svg path {
    fill: var(--danger);
}

form.inline-form {
    display: flex;
    margin-top: 40px;
    gap: 8px;
}

input.input__text {
    align-items: center;
    background: var(--gray-500);
    border-radius: 8px;
    border: 1px solid var(--gray-700);
    color: var(--gray-300);
    /* display: flex; */
    flex: 1 0 0;
    font-family: Inter;
    font-size: 16px;
    font-style: normal;
    font-weight: 400;
    gap: 8px;
    line-height: 140%;
    padding: 16px;
}

input.input__text:hover,
input.input__text:active,
input.input__text:focus-visible {
    border: 1px solid var(--purple-dark);
    color: var(--gray-100);
    outline: none;
}

/* Change the white to any color */
input.input__text:-webkit-autofill,
input.input__text:-webkit-autofill:hover,
input.input__text:-webkit-autofill:focus,
input.input__text:-webkit-autofill:active {
    -webkit-box-shadow: 0 0 0 30px var(--gray-500) inset !important;
}



input[type='checkbox'].input__checkbox {
    -webkit-appearance: none;
    appearance: none;
    background-color: transparent;
    cursor: pointer;
    height: 28px;
    margin: 0;
    width: 28px;
}

input[type='checkbox'].input__checkbox {
    content: url('./assets/circle-unchecked.svg');
}

input[type='checkbox'].input__checkbox:checked {
    content: url('./assets/circle-checked.svg');
}

input[type='checkbox'].input__checkbox,
input[type='checkbox'].input__checkbox:checked {
    min-height: 18px;
    min-width: 18px;
    padding: 5px;
}

input[type='checkbox'].input__checkbox:active,
input[type='checkbox'].input__checkbox:focus-visible {
    background: var(--gray-400);
    border-radius: 4px;
    outline: none;
}

label.card__description {
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 2;
    color: var(--gray-100);
    cursor: pointer;
    display: -webkit-box;
    font-family: Inter;
    font-size: 14px;
    font-style: normal;
    font-weight: 400;
    line-height: 140%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: initial;
}

label.card__description--completed {
    color: var(--gray-300);
    font-family: Inter;
    font-size: 14px;
    font-style: normal;
    font-weight: 400;
    line-height: 140%;
    /* 19.6px */
    text-decoration-line: line-through;
}



/* List */

.todo__list {
    flex-direction: column;
    margin-block: 25px;
    row-gap: 12px;
}

/* Card */

.todo__card {
    align-items: center;
    align-self: stretch;
    background: var(--gray-500);
    border-radius: 8px;
    border: 1px solid var(--gray-400);
    box-shadow: 0px 2px 8px 0px rgba(0, 0, 0, 0.06);
    display: flex;
    gap: 12px;
    height: 100%;
    justify-content: space-between;
    max-height: 72px;
    padding-block: 16px;
    padding-inline: 11px 16px;
}

.todo__card .card__content {
    cursor: pointer;
    gap: 7px;
}

/* TodoHeader */
.count__container {
    gap: 8px;
    width: fit-content;
}

.count {
    background-color: var(--gray-400);
    border-radius: 999px;
    color: var(--gray-200);
    padding: 2px 8px;
    text-align: center;
    width: max-content;
}


/* TodoEmpty */

.todo-empty__container {
    border-top: 1px solid var(--gray-400);
    display: flex;
    flex-direction: column;
    gap: 16px;
    margin-block: 24px;
    padding-block: 64px;
}

.todo-list__container {
    margin-block: 24px;
}