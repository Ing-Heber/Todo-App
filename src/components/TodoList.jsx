import PropTypes from 'prop-types'
import { TodoItem } from "./TodoItem"

export const TodoList = ({ todos, toggleTodo, deleteTodo }) => {
    return (
        <div className="todo__list d-flex flex-column">
            {
                todos && todos.map(({ id, description, completed }) => (
                    <TodoItem
                        key={id}
                        id={id}
                        description={description}
                        completed={completed}
                        toggleTodo={toggleTodo}
                        deleteTodo={deleteTodo}
                    />
                ))
            }
        </div>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array,
    toggleTodo: PropTypes.func,
    deleteTodo: PropTypes.func
}