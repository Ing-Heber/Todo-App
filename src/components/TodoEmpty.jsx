import './styles.css';
import PropTypes from "prop-types";

export const TodoEmpty = ({pendingTodos = 0}) => {
    const count = Number.isFinite(pendingTodos) ? pendingTodos : 0;
    const isPending = count > 0;
    const plural = count === 1 ? 'task' : 'tasks';

    const titles = {
        pending: 'Carry On!',
        completed: 'Congratulations!'
    };
    const messages = {
        pending: `You have ${count} ${plural} to finish`,
        completed: 'You have finished your pendings'
    };

    const text = isPending
        ? {title: titles.pending, message: messages.pending}
        : {title: titles.completed, message: messages.completed};

    return (
        <div className='todo-empty__container d-flex flex-column align-items-center animate__animated animate__fadeIn'>
            <img src='/assets/clipboard.svg' alt="Clipboard" width={56} />
            <h3 className='text-center color-gray-300'>
                {text.title} <br/>
                <span className='fw-regular'>{text.message}</span>
            </h3>
        </div>
    );
};

TodoEmpty.propTypes = {
    pendingTodos: PropTypes.number
};