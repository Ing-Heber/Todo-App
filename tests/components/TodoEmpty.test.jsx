import { render, screen } from "@testing-library/react"
import { TodoEmpty } from "../../src/components/TodoEmpty"

describe('Testing <TodoEmpty />', () => { 

    test('should match with snapshot', () => { 
        const {container} = render(<TodoEmpty />)
        expect(container).toMatchSnapshot();
     })

     test('should render image', () => { 
        render(<TodoEmpty />)
        expect(screen.getByAltText('Clipboard image')).toBeTruthy()
     })

     test('should render header', () => { 
        render(<TodoEmpty />)
        expect(screen.getByRole('heading', {name: 'Congratulations! You have finished your pendings',level: 3})).toBeTruthy();
     })
    

 })