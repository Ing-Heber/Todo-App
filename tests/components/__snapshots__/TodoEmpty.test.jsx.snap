// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <TodoEmpty /> should match with snapshot 1`] = `
<div>
  <div
    class="todo-empty__container d-flex flex-column align-items-center animate__animated animate__fadeIn"
  >
    <img
      alt="Clipboard image"
      height="auto"
      src="/assets/clipboard.svg"
      width="56px"
    />
    <h3
      class="text-center color-gray-300"
    >
      Congratulations! 
      <br />
      <span
        class="fw-regular"
      >
         You have finished your pendings
      </span>
    </h3>
  </div>
</div>
`;
